src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java

```diff
@@ -738,8 +735,9 @@ public void beforeCommand(CommandEnvironment env) throws AbruptExitException {
               digestUtil.getDigestFunction(),
               remoteOptions,
               verboseFailures,
-              fallbackDownloader));
+              fallbackDownloader);
       downloaderChannel.release();
+      env.getDownloaderDelegate().setDelegate(remoteDownloader);
```

Comment:
Would this break repository download?

Reply:
What do you mean by repository download?

The way it's supposed to work (and how I understood the previous code path too, that was just more involved):
* The `DelegatingDownloader` is initialized with a `HttpDownloader`
* Consumers pull in the `DelegatingDownloader` object when they're initialized.
* Once the `RemoteModule` is initialized, if a remote (grpc) downloader is configured, that is set into `DelegatingDownloader`. If that object is non-null, it's used now instead of the `HttpDownloader` for all downloads.

Reply:
I ran some tests, looks like it does. But it is already the case before this change with [remoteDownloaderSupplier](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java;drc=38ade70fed4b34043d21a606717ad2ed1ee9c696;bpv=1;bpt=1;l=168).

However this behaviour does feel strange to me: `--experimental_remote_downloader` can only be used with `--remote_cache` and somehow it affects repository download (`ctx.download_and_extract`) as well.

I got this error with both Bazel 7.2.1 and Bazel built from this PR
```
bazel build --experimental_remote_downloader=grpc://127.0.0.1:9092 --remote_cache=grpc://127.0.0.1:9092 //examples/cpp:hello-world --repository_cache=
...
ERROR: An error occurred during the fetch of repository 'platforms':
   Traceback (most recent call last):
	File "/usr/local/google/home/pcloudy/.cache/bazel/_bazel_pcloudy/3d42a5c7654876aa68e91f3b969deb88/external/bazel_tools/tools/build_defs/repo/http.bzl", line 131, column 45, in _http_archive_impl
		download_info = ctx.download_and_extract(
Error in download_and_extract: java.io.IOException: io.grpc.StatusRuntimeException: UNIMPLEMENTED: unknown service build.bazel.remote.asset.v1.Fetch
ERROR: no such package '@@platforms//host': java.io.IOException: io.grpc.StatusRuntimeException: UNIMPLEMENTED: unknown service build.bazel.remote.asset.v1.Fetch
ERROR: /usr/local/google/home/pcloudy/.cache/bazel/_bazel_pcloudy/3d42a5c7654876aa68e91f3b969deb88/external/bazel_tools/tools/BUILD:97:6: @@bazel_tools//tools:host_platform depends on @@platforms//host:host in repository @@platforms which failed to fetch. no such package '@@platforms//host': java.io.IOException: io.grpc.StatusRuntimeException: UNIMPLEMENTED: unknown service build.bazel.remote.asset.v1.Fetch
ERROR: /usr/local/google/home/pcloudy/workspace/my_tests/bazel/examples/cpp/BUILD:7:10: Target @@bazel_tools//tools:host_platform was referenced as a platform, but does not provide PlatformInfo
ERROR: Analysis of target '//examples/cpp:hello-world' failed; build aborted
```

Reply:
LGTM since there is no regression, but we probably do want to make sure `ctx.download_and_extract` work with `--experimental_remote_downloader` enabled?

Reply:
I don't quite understand the error you're getting - agianst which service are you running there?

Given the documentation ("A Remote Asset API endpoint URI, to be used as a remote download proxy"), redirecting `ctx.download*` is exactly what I'd expect `GrpcRemoteDownloader` to do. I don't think it's related to the remote cache per se, except that a cache service can also choose to be a downloader service (https://github.com/buchgr/bazel-remote for example is).
That downloader service then (presumably) caches the downloads it's seen, and lives closer on the network than the upstream - then you save on traffic and latency across the internet. 

> UNIMPLEMENTED: unknown service build.bazel.remote.asset.v1.Fetch

seems like your remote cache you're running locally doesn't implement the remote download api?

Reply:
I was testing with https://github.com/buchgr/bazel-remote, is it possible that ctx.download_and_extract no longer works for downloading http URLs when using the GrpcRemoteDownloader?

Reply:
Could it be that you didn't set `--experimental_remote_asset_api` for your bazel-remote invocation? That way, I can reproduce the error you've got, with it, everything works as expected.
I'd also assume there's an integration test ensuring this behavior doesn't break.

Reply:
I see, I was missing `--experimental_remote_asset_api` flag for `bazel-remote`, after adding that flag, everything works! So everything is WAI! Sorry for the noise.